### ISSUE TITLE
JupyterLite demo

### ISSUE THREAD

**Issue by juntyr** [ecmwf/earthkit#41]
**TLDR:** Try earthkit in your browser without installing anything [here](https://esiwace3-compression-lab.onrender.com/notebooks?fromURL=https://gist.githubusercontent.com/juntyr/122a8868765d48ca993ac7df3473e621/raw/earthkit.ipynb)

I'm a PhD student working on the online data compression laboratory for ESiWACE3. After I first heard about EarthKit yesterday, I thought I could try to add it to my online lab. In short, the online lab uses a custom Pyodide build (Python running inside your web browser using WebAssembly) together with JupyterLite (JupyterLab but using Pyodide) and adds support for several meteorology-specific Python packages that are not yet (publicly) supported for WebAssembly. The key selling point of the lab is that it is immediately accessible to anyone without them needing to install anything. As work on the online lab and its connected publication is ongoing, some of the magic behind the scenes is not open-source quite yet, but I wanted to share the small demo with you.

If you'd be interested in hosting a larger online demo for earthkit, e.g. to provide runnable documentation examples, I'd be happy to help you setup a similar Pyodide build.

**Comment by tlmquintino** [comment]
thanks for this. it is really interesting. I get an error while running the code in the link, but generally speaking I think we would be interested in this, especially to have some runnable earthkit demos.
Please do get in touch with us.

**Comment by juntyr** [comment]
You are asking about the magic behind the scenes :)

For Pyodide, all dependencies are compiled to WebAssembly. For pure Python packages, nothing has to be done. To get eccodes working, I compile libeccodes to WebAssembly as a shared library (note that WASM32 is a 32bit system, eccodes only supports 64bits, but disabling its check for that "just works") and slightly patch the `gribapi/bindings.py` to find the library (and eccodes definitions) in an unusual path.

At the moment, the Pyodide project manually maintains a huge variety of non-pure Python packages to make sure that they can be used inside the browser. In my online lab, I have simply extended this collection with several more packages. In the long run, PyPi will add support for developers to publish WASM-compiled versions there directly and WebAssembly will just become another platform that Python packages may or may not support.

If a demo such as this one works, WebAssembly support could be upstreamed back into e.g. eccodes and I could try to get it into the roster of Pyodide-supported packages (e.g. so that anyone using the "Try Jupyter" page, which also uses JupyterLite, could just import eccodes there).

**Comment by MMesch** [comment]
Hi all,

I work at QuantStack. We are authors and maintainers of JupyterLite and the emerging [emscripten-forge](https://github.com/emscripten-forge) ecosystem which is basically a conda-channel for WASM packages.

I also tried what @juntyr did. It's running and you can test it yourself under https://notebook.link/@mmesch/earthkit-test  (Notebook.link is a new Jupyter platform and can be used to share JupyterLite environments). Some notes and observations:

- earthkit is pulled directly from pypi and we only need to apply a small monkey patch run-time because of os.link usage that doesn't work well in WASM, and map reprojections in another package that are parallel by default.
- Emscripten-forge is used for compiled dependencies such as hdf5/netcdf, numpy
- The Copernicus CDS is [**CORS blocked**](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CORS/Errors) which is really the main pain point when using earthkit from the browser. I used a manual CORS proxy to circumvent this locally. Basically it downloads files locally and then passes things forward into the browser via custom declarative net rules. If this is done, download works perfectly fine from the browser.
- I did not yet compile libraries for GRIB, codecs etc for WASM yet. Judging from the nice work of @juntyr and also some preliminary exploration I did, it seems this is doable without hard blockers. But since I am not directly working in this field, I thought it's best to check what would actually makes sense and whether there is interest to have this available in browser.
- although earthkit runs quite acceptable in-browser for the type of analysis shown, further performance optimizations are likely possible (e.g. faster file reads).


I find it's amazing that anyone can use earthkit to do simple calculations without any installation and without any backend from ECMWF other than the data store, directly in the browser. It's particularly fun with an AI agent run-time helping to identify data, writing the actual download, plotting, transforming and analysing. It's also quite powerful in combination with other WASM-enabled libraries such as Geopandas because that gives way more capability than a classic website ever could.  To me, CORS is the major blocker to this use case and I'd be curious to hear arguments whether it could be enabled or not.

Please let me know if you have any questions or want to discuss further. Also CCing @SylvainCorlay from our end.

---

libraries available in this environments (I also have the lock file with specific versions in case that's of interest):

```
name: default-environment
channels:
  - https://prefix.dev/emscripten-forge-4x
  - https://prefix.dev/conda-forge
dependencies:
  - xeus-python
  - xarray
  - geopandas
  - jupytergis-lite
  - rioxarray
  - zarr   # doesn't work well yet but could (https://github.com/emscripten-forge/recipes/pull/4228)
  - netCDF4
  - numcodecs
  - gcsfs
  - pyyaml
  - cartopy
  - scipy
  - pip:
    - cdsapi
    - earthkit
    - earthkit-maps
    - earthkit-data
```

**Comment by juntyr** [comment]
Thanks @MMesch for your investigation! I also wanted to give a quick update: I have been working on ESiWACE3's Online Laboratory for Climate Science and Meteorology (https://lab.climet.eu/) for three years now. It's JupyterLite with some extra ported packages like cfgrib and also some custom usability features, e.g. automatically working around CORS issues. earthkit has been running in the lab for most of this time. I have also now used this more and more for interactive online documentation in my own projects, and still think that it would be very valuable to have, especially if more WASM patches can be upstreamed, particularly now that one can publish PyEmscripten wheels to PyPi.

**Comment by MMesch** [comment]
Nice, very cool that you are using such a deployment already @juntyr . Happy to sync on this more if you like. I think we could get to a point where this stack works pretty much out of the box. I only encountered the os.link issue and the projection serialization issues requiring upstreaming for now, but mostly used the netcdf code path. Pushing more of the compiled packages to emscripten-forge is quite fast. I also wanted to think about secret management more to make sure it's safe. But CORS is the main blocker for me because workarounds are all not great.

**Comment by MMesch** [comment]
Also, maybe making the earthkit documentation interactive could be a nice short term objective. It would make sure that everything runs well upstream.
